src/target/riscv/riscv-013.c

```diff
@@ -1911,6 +1927,135 @@ static int set_group(struct target *target, bool *supported, unsigned int group,
 	return ERROR_OK;
 }
 
+static int wait_for_idle_if_needed(struct target *target)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
+		return ERROR_FAIL;
+	if (!dm->abtractcs_busy)
+		return ERROR_OK;
+
+	uint32_t abstractcs;
+	int result = wait_for_idle(target, &abstractcs);
+	if (result != ERROR_OK)
+		return result;
+	LOG_DEBUG_REG(target, DM_ABSTRACTCS, abstractcs);
+	return ERROR_OK;
+}
+
+static int examine_dm(struct target *target)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
+		return ERROR_FAIL;
+	if (dm->was_examined)
+		return ERROR_OK;
+
+	int result = ERROR_FAIL;
+
+	uint32_t dmcontrol;
+	if (!dm->was_reset) {
+		/* First, the Debug Module is reset. However,
+		 * `dmcontrol.hartsel` should be read first, in order not to
+		 * change it when requesting the reset, since changing it
+		 * without checking that `abstractcs.busy` is low is
+		 * prohibited.
+		 */
+		result = dm_read(target, &dmcontrol, DM_DMCONTROL);
+		if (result != ERROR_OK)
+			return result;
+		/* Initiate the reset (`dmcontrol.dmactive == 0`) leaving
+		 * `dmcontrol.hartsel` the same.
+		 */
+		dmcontrol = (dmcontrol & DM_DMCONTROL_HARTSELLO) |
+			(dmcontrol & DM_DMCONTROL_HARTSELHI);
+		result = dm_write(target, DM_DMCONTROL, dmcontrol);
+		if (result != ERROR_OK)
+			return result;
+	} else {
+		/* If the DM was not reset, an abstract command can be executed
+		 * at the moment.
+		 */
```

Comment:
I have tried to update the wording for clarity

```suggestion
		/* If the DM was already reset (when examining a different hart).
		 * No need to reset it again. But for safety, assume that an abstract 
		 * command might be in progress at the moment.
		 */
```

Reply:
Thanks! Addressed (the wording is a bit different, please check).